src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  const newState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    }
 
+    if (action.type === 'clear') {
+      for (const key in newState) {
+        delete newState[key];
+      }
+    }
+
+    stateHistory.push({ ...newState });
+  }
+
+  return stateHistory;
+}
```

Comment:
Remember to include a `default` case in the `switch` statement for error handling. This will catch any unexpected `action.type` values and can be used to throw an error or handle the case appropriately.